stdlib/public/core/Diffing.swift

```diff
@@ -0,0 +1,805 @@
+  >(
+    from source: Source,
+    to target: Target,
+    by areEquivalent: (Source.Element, Target.Element) -> Bool
+  ) where
+    Source.Element == Target.Element,
+    Source.Index == SourceIndex,
+    Target.Index == TargetIndex
+  {
+    let pathStart = (x: source.startIndex, y: target.startIndex)
+    let pathEnd = (x: source.endIndex, y: target.endIndex)
+    let matches = source._commonPrefix(with: target, by: areEquivalent)
+    let (x, y) = (matches.0.endIndex, matches.1.endIndex)
+
+    if pathStart == pathEnd {
+      pathStorage.removeAll(keepingCapacity: true)
+      pathStartIndex = 0
+    } else if x == pathEnd.x || y == pathEnd.y {
+      pathStorage.removeAll(keepingCapacity: true)
+      pathStorage.append(pathStart)
+      if pathStart != (x, y) && pathEnd != (x, y) {
+        pathStorage.append((x, y))
+      }
+      pathStorage.append(pathEnd)
+      pathStartIndex = 0
+    } else {
+      formChangesCore(from: source, to: target, x: x, y: y, by: areEquivalent)
+    }
+  }
+
+  /// The core difference algorithm.
+  ///
+  /// - Precondition: There is at least one difference between `a` and `b`
+  /// - Runtime: O(*n* * *d*), where *n* is `a.count + b.count` and
+  ///   *d* is the number of inserts and removes.
+  /// - Space: O(*d* * *d*), where *d* is the number of inserts and removes.
+  private mutating func formChangesCore<
+    Source: BidirectionalCollection,
+    Target: BidirectionalCollection
+  >(
+    from a: Source,
+    to b: Target,
+    x: Source.Index,
+    y: Target.Index,
+    by areEquivalent: (Source.Element, Target.Element) -> Bool
+  ) where
+    Source.Element == Target.Element,
+    Source.Index == SourceIndex,
+    Target.Index == TargetIndex
+  {
+    // Written to correspond, as closely as possible, to the psuedocode in
+    // Myers, E. "An O(ND) Difference Algorithm and Its Variations".
+    //
+    // See "FIGURE 2: The Greedy LCS/SES Algorithm" on p. 6 of the [paper].
+    //
+    // Note the following differences from the psuedocode in FIGURE 2:
+    //
+    // 1. FIGURE 2 relies on both *A* and *B* being Arrays. In a generic
+    //  context, it isn't true that *y = x - k*, as *x*, *y*, *k* could
+    //  all be different types, so we store both *x* and *y* in *V*.
+    // 2. FIGURE 2 only reports the length of the LCS/SES. Reporting a
+    //  solution path requires storing a copy of *V* (the search frontier)
+    //  after each iteration of the outer loop.
+    // 3. FIGURE 2 stops the search after *MAX* iterations. We run the loop
+    //  until a solution is found. We also guard against incrementing past
+    //  the end of *A* and *B*, both to satisfy the termination condition
+    //  and because that would violate preconditions on collection.
+    //
+    // [paper]: http://www.xmailserver.org/diff2.pdf
+    var (x, y) = (x, y)
+    let (n, m) = (a.endIndex, b.endIndex)
+
+    var v = _SearchState<Source.Index, Target.Index>(consuming: &pathStorage)
```

Comment:
Vrbl nms R nt vry rdbl.

Reply:
Sorry, too used to writing pure C—vowels are expensive!

Reply:
I think the terse names are justified here since the implementation follows the paper, and using the same notation makes the code easier to follow.